### ISSUE TITLE
TypeError: Tensors in list passed to 'values' of 'ConcatV2' Op have types [int8, float32, float32] that don't all match.

### ISSUE THREAD

**Issue by szgabsz91** [robertostling/sigmorphon2016-system#1]
Hi Robert,

I tried to test this system but couldn't manage to execute the commands mentioned in README.md.

* OS: macOS 10.14.5
* Python: 3.7.3
* Pip: 19.0.3
* Keras: 2.1.6 (I downgraded because in newer versions, the merge function has been removed)
* Tensorflow: 1.13.1

Log:

```
Using TensorFlow backend.
sigmorphon-conv.py:69: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(32, 128, input_length=38, embeddings_constraint=<keras.con...)`
  W_constraint=maxnorm(2))
WARNING:tensorflow:From /usr/local/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
sigmorphon-conv.py:76: UserWarning: The `merge` function is deprecated and will be removed after 08/2017. Use instead layers from `keras.layers.merge`, e.g. `add`, `concatenate`, etc.
  mode='concat')
/usr/local/lib/python3.7/site-packages/keras/legacy/layers.py:465: UserWarning: The `Merge` layer is deprecated and will be removed after 08/2017. Use instead layers from `keras.layers.merge`, e.g. `add`, `concatenate`, etc.
  name=name)
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 455, in _apply_op_helper
    as_ref=input_arg.is_ref)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1240, in internal_convert_n_to_tensor
    ctx=ctx))
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1175, in internal_convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 977, in _TensorTensorConversionFunction
    (dtype.name, t.dtype.name, str(t)))
ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int8: 'Tensor("repeat_vector_1/Tile:0", shape=(?, 38, 25), dtype=int8)'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "sigmorphon-conv.py", line 416, in <module>
    [x for x in sys.argv[2:] if x != 'TESTING'])
  File "sigmorphon-conv.py", line 406, in run_experiments
    acc = train_model(prefix, '%s/%s' % (data_path, language), **c)
  File "sigmorphon-conv.py", line 309, in train_model
    morphon.build(**kwargs)
  File "sigmorphon-conv.py", line 76, in build
    mode='concat')
  File "/usr/local/lib/python3.7/site-packages/keras/legacy/layers.py", line 465, in merge
    name=name)
  File "/usr/local/lib/python3.7/site-packages/keras/legacy/layers.py", line 128, in __init__
    self(input_tensors, mask=input_masks)
  File "/usr/local/lib/python3.7/site-packages/keras/engine/topology.py", line 619, in __call__
    output = self.call(inputs, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/keras/legacy/layers.py", line 220, in call
    return K.concatenate(inputs, axis=self.concat_axis)
  File "/usr/local/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py", line 1885, in concatenate
    return tf.concat([to_dense(x) for x in tensors], axis)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py", line 180, in wrapper
    return target(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py", line 1256, in concat
    return gen_array_ops.concat_v2(values=values, axis=axis, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1149, in concat_v2
    "ConcatV2", values=values, axis=axis, name=name)
  File "/usr/local/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 483, in _apply_op_helper
    raise TypeError("%s that don't all match." % prefix)
TypeError: Tensors in list passed to 'values' of 'ConcatV2' Op have types [float32, int8] that don't all match.
```

Could you please help in what went wrong? I guess I am using wrong library versions, but I'm not sure what versions should I use.

Thanks in advance!

**Comment by robertostling** [comment]
Hi Gabor,

I don't remember exactly which version of Keras I used, but it should be from early 2016, so before version 1.0.0 even. And back then, I used the Theano backend (not TensorFlow). Not sure how painful that would be to install today, deep learning certainly changes a lot in 3 years...

That said, it looks like my code is wrong, and maybe Keras finally caught this. I'm actually trying to concatenate a float32 and an int8 tensor like the error message says, and I assume the old version did an implicit conversion. Please try the following:

1. change from int8 to float32 on line 56 of sigmorphon-conv.py
2. change from int8 to float32 on line 140 of data.py

I hope this is everything. If you test this and it works, please tell me and I'll commit the changes (or make a pull request).

Robert

**Comment by szgabsz91** [comment]
Thanks for your help. I managed to test the system with the Tensorflow backend and now it completed successfully (although in CPU mode) with your suggestions.

I also updated the other Python file, and added a `requirements.txt` to make sure that the dependencies can be reconstructed later. Plus I updated `README.md` to include information about how to use virtualenv and how to automatically install these dependencies.

See my pull request #2.

Thanks again!
